crates/solana-gateway-task-processor/src/component.rs

```diff
@@ -398,6 +404,20 @@ async fn execute_task(
     Ok(())
 }
 
+/// Checks if the incoming message has already been executed.
+async fn incoming_message_already_executed(
+    solana_rpc_client: &RpcClient,
+    incoming_message_pda: &Pubkey,
+) -> eyre::Result<bool> {
+    let raw_incoming_message = solana_rpc_client
+        .get_account_data(incoming_message_pda)
+        .await?;
+    let incoming_message = IncomingMessage::read(&raw_incoming_message)
+        .ok_or_eyre("failed to read incoming message")?;
+
+    Ok(incoming_message.status == MessageStatus::Executed)
```

Comment:
I just wonder if this would be a relatively expensive operation (One io RPC call) and how far we would be from implementing an [LRU](https://docs.rs/lru/latest/lru/) cache. I guess this need all depends how many "hits" we are expecting here. If we are hitting this check all the time, i would maybe go for it (on a separate issue). Wdyt?

Reply:
In the standard flow we'd get 1 hit per 1 message upon relayer startup, no extra calls. I don't think LRU cache would help us here.

This is because the relayer gets "tasks" from the Amplifier API that are always sequential and always unique. This means that there's only ever gonna be a single "execute GMP message" task for a single message incoming message (unless the relayer gets restarted and we query the same tasks from the Amplifier API -- but even then LRU cache would not help us here)

Reply:
Yeah, I'm not sure how caching would help us here. I mean, it would be nice to be able to avoid this call, but we would likely have a cache miss most of the time and thus the cost of checking the cache would not be worth it.

Reply:
Okay i see. So the amplifier API only works with "exactly once" and not with "at least once" semantics. Thanks !